pkg/referrerstore/oras/oraswithcache.go

```diff
@@ -0,0 +1,133 @@
+package oras
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/deislabs/ratify/pkg/common"
+	"github.com/deislabs/ratify/pkg/ocispecs"
+	"github.com/deislabs/ratify/pkg/referrerstore"
+	"github.com/deislabs/ratify/pkg/referrerstore/config"
+	"github.com/dgraph-io/ristretto"
+	"github.com/dgraph-io/ristretto/z"
+	"github.com/opencontainers/go-digest"
+	"github.com/sirupsen/logrus"
+)
+
+type Operation int
+
+const (
+	Name Operation = iota
+	GetConfig
+	ListReferrers
+	GetBlobContent
+	GetReferenceManifest
+	GetSubjectDescriptor
+)
+
+type orasStoreWithInMemoryCache struct {
+	base      referrerstore.ReferrerStore
+	cache     *ristretto.Cache
+	cacheConf *OrasCacheConf
+}
+
+type OrasCacheConf struct {
+	Enabled bool `json:"cacheEnabled"`
+	Ttl     int  `json:"ttl"`
+}
+
+type orasStoreFactoryWithCache struct{}
+
+func keyToHash(key interface{}) (uint64, uint64) {
+	if key == nil {
+		return 0, 0
+	}
+	switch k := key.(type) {
+	case string:
+		return z.MemHashString(k), xxhash.Sum64String(k)
+	default:
+		return 0, 0
+	}
+}
+
+// Create creates a new oras store decorated with in-memory cache to cache
+// results of ListReferrers API.
+func (s *orasStoreFactoryWithCache) Create(storeBase referrerstore.ReferrerStore, cacheConf *OrasCacheConf) (referrerstore.ReferrerStore, error) {
+	memoryCache, err := ristretto.NewCache(&ristretto.Config{
```

Comment:
Have you looked at the caches available in [api-machinery](https://github.com/kubernetes/apimachinery/tree/master/pkg/util/cache)? Curios if you got a chance to try those as well.

Reply:
Also, it might be good to add some details in the PR description and/or comments in the code stating why this cache package was chosen

Reply:
Thanks for calling it out! I'll give it a try! Once I make the decision, I'll update the description.

Reply:
The cache under [apimachinery](https://github.com/kubernetes/apimachinery) does not have sharding on map, which might not be good on concurrency efficiency.